File: 1032-stream-of-characters/1032-stream-of-characters.py
class TrieNode:
    def __init__(self):
        self.next = {}
        self.is_word = False
        

class StreamChecker:

    def __init__(self, words: List[str]):
        self.s = ''
        self.root = TrieNode()
        
        for word in words:
            self.insert(word[::-1])
            

    def query(self, letter: str) -> bool:
        self.s = letter + self.s
        node = self.root
        for ch in self.s:
            if ch not in node.next:
                return False
            node = node.next[ch]
            if node and node.is_word:
                return True
        return False
        
        
    def insert(self, word):
        node = self.root
        for ch in word:
            node = node.next.setdefault(ch, TrieNode())
        node.is_word = True


# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)

File: 1032-stream-of-characters/1032-stream-of-characters.js
class TrieNode {
  constructor() {
    this.next = new Array(26).fill(null);
    this.isWord = false;
  }
}

/**
 * @param {string[]} words
 */
var StreamChecker = function(words) {
  this.word = '';
  this.root = new TrieNode();
  
  for (let word of words) {
    let node = this.root;
    for (let i = word.length - 1; i >= 0; i--) {
      const index = word.charCodeAt(i) - 'a'.charCodeAt();
      if (!node.next[index]) node.next[index] = new TrieNode();
      
      node = node.next[index];
    }
    
    node.isWord = true;
  }

};

/** 
 * @param {character} letter
 * @return {boolean}
 */
StreamChecker.prototype.query = function(letter) {
  this.word = letter + this.word;
  let node = this.root;
  
  for (let ch of this.word) {
    const index = ch.charCodeAt() - 'a'.charCodeAt();
    if (!node.next[index]) return false;
    
    node = node.next[index];
    if (node && node.isWord) return true;
  }
  
  return false;
};

/** 
 * Your StreamChecker object will be instantiated and called as such:
 * var obj = new StreamChecker(words)
 * var param_1 = obj.query(letter)
 */